### PR TITLE
Enumerate pros/cons of explicit/implicit vector length

### PR DESCRIPTION
**Pull request by rofirrim**
I'm trying to better enumerate the pros and cons of the two approaches.

### CODE REVIEW

File: rvv-intrinsic-rfc.md [hunk 1]
```diff
@@ -308,21 +308,27 @@ There are two variants of intrinsics regarding to `vl`.
...
 
-    Pass `vl` as an argument to the intrinsics. The `vl` argument helps users to track `vl` values from previous `vsetvl`.  It is strict about using `vl` only coming from an earlier `vsetvl` or `vsetvlmax` intrinsics. It inhibits to do arithmetic on `vl` argument.
+    All the intrinsics receive an explicit `vl` argument, by value, which is used by the vector operation. Only calls to `vsetvl` or `vsetvlmax` intrinsics can be used to generate values that are permissible as the explicit `vl` argument.
 
-    The design philosophy of explicit vl intrinsics is
-     - Operations could be entirely defined by the operands.
-     - In C programmers' point of view, there is no side-effect of intrinsics with `vl` argument.
+    Pros:
+     - Operations are entirely defined by the vector operands and the explicit vector length.
+     - In C programmers' point of view, there is no side-effect of intrinsics with `vl` argument. So there is no need to give specific semantics to the global nature of `vl` in a C program.
```
[the remark this answers is not in the document]
Comment (rofirrim) [reply rvv-intrinsic-rfc.md:315]: I see. Then it means _implicit vl operand_ at the eyes of a C programmer is not like a global variable but instead it acts more like an implicit local variable (or parameter) initialized with the value that the `vl` register had upon entry the function. 

Note the two concepts here: we can only change the _implicit vl operand_ in C (via the vsetvl or other vl-setting operations) of the _current function_, regardless that the actual storage is the `vl` register.

Does this align with your expectation?

File: rvv-intrinsic-rfc.md [hunk 2]
```diff
@@ -308,21 +308,27 @@ There are two variants of intrinsics regarding to `vl`.
...
+
+    Cons:
+     - It is less consistent with the C operator syntax. The GCC vector extension cannot be implemented, because of lack of the `vl` operand, or it must be given whole vector register semantics.
+     - It is more verbose, as many vector codes may not use different `vl` values at the same time.
 
 2. Implicit vl intrinsics:
 
-    No `vl` argument. Users need to know where the `vl` comes by the program execution flow. Users need to take care by themselves if they want to mix multiple `vl` operations.
+    Intrinsics do not receive a `vl` argument. Instead the `vl` used by the vector operations is the one set by the last call to `vsetvl` or `vsetvlmax` intrinsics that the program has done at runtime. The user is responsible for setting the `vl`.
```
[the remark this answers is not in the document]
Comment (rofirrim) [reply rvv-intrinsic-rfc.md:323]: Sure I will do that.

File: rvv-intrinsic-rfc.md [hunk 3]
```diff
@@ -308,21 +308,27 @@ There are two variants of intrinsics regarding to `vl`.
...
+     - It is more consistent with C operator syntax as vector operations naturally involve the implicit `vl` argument.
 
-    The design philosophy of implicit vl intrinsics is
-     - The user is responsible for setting the `vl`.
-     - It is more faithful to the underlying assembly code.
-     - It seems to open the possibility for additional programming errors through additional vl argument.
-     - It is more consistent with C operator syntax.
+    Cons:
+     - The implicit `vl` operand might open the possibility for unobvious programming errors.
+     - May force deciding upfront whether programmer's view of `vl` is a global variable or an implicit parameter of the function.
```
[the remark this answers is not in the document]
Comment (rofirrim) [reply rvv-intrinsic-rfc.md:331]: Ok. Fair enough.

File: rvv-intrinsic-rfc.md [hunk 4]
```diff
@@ -308,21 +308,27 @@ There are two variants of intrinsics regarding to `vl`.
...
-    The design philosophy of explicit vl intrinsics is
-     - Operations could be entirely defined by the operands.
-     - In C programmers' point of view, there is no side-effect of intrinsics with `vl` argument.
+    Pros:
+     - Operations are entirely defined by the vector operands and the explicit vector length.
+     - In C programmers' point of view, there is no side-effect of intrinsics with `vl` argument. So there is no need to give specific semantics to the global nature of `vl` in a C program.
+
+    Cons:
+     - It is less consistent with the C operator syntax. The GCC vector extension cannot be implemented, because of lack of the `vl` operand, or it must be given whole vector register semantics.
+     - It is more verbose, as many vector codes may not use different `vl` values at the same time.
```
[the remark this answers is not in the document]
Comment (rofirrim) [reply rvv-intrinsic-rfc.md:319]: Ok!

**Comment by zakk0610** [comment]
IMPO, listing pros/cons will be very helpful, what's stuck this PR?

BTW, I think the another pros for explicit vector length is "It is very clear for users to know the operation is regard or regardless of vector length". 
For example, `vmv.x.s` and `vfmv.f.s` always copy element 0 (regardless of vl), but it's ambiguous in implicit vl interface, the document need to have more clarification for the relation of operation and vector length.

**Comment by rofirrim** [comment]
Hi, sorry it got stuck because I never got round to address Kai's comments. I'll try to find some time for this.

**Comment by rofirrim** [comment]
> For example, `vmv.x.s` and `vfmv.f.s` always copy element 0 (regardless of vl), but it's ambiguous in implicit vl interface, the document need to have more clarification for the relation of operation and vector length.

But these two intrinsics would never receive an explicit vl, would they?

**Comment by zakk0610** [comment]
> Hi, sorry it got stuck because I never got round to address Kai's comments. I'll try to find some time for this.

> > For example, `vmv.x.s` and `vfmv.f.s` always copy element 0 (regardless of vl), but it's ambiguous in implicit vl interface, the document need to have more clarification for the relation of operation and vector length.
> 
> But these two intrinsics would never receive an explicit vl, would they?

Yes. 
BTW current document is wrong in rvv_intrinsic_funcs_vl interface which has `vl` argument( ex. `vmv_x_s_i8mf8_i8_vl (vint8mf8_t src, size_t vl`).

**Comment by rofirrim** [comment]
Maybe I don't understand github but I pushed new commits in https://github.com/rofirrim/rvv-intrinsic-doc/tree/rofirrim_pros_cons_explicit_implicit_vl

Not sure if there's a way to force an update.

**Review by Hsiangkai — DISMISSED** [review]

**Comment by Hsiangkai** [comment]
> Maybe I don't understand github but I pushed new commits in https://github.com/rofirrim/rvv-intrinsic-doc/tree/rofirrim_pros_cons_explicit_implicit_vl
> 
> Not sure if there's a way to force an update.

It is strange. I could not see your the other two commits in your branch in this pull request. I am also unfamiliar with github.

**Comment by rofirrim** [comment]
Let's do a simpler thing: let's close it and I'll reopen another PR.

**Comment by rofirrim** [comment]
Continue at https://github.com/riscv/rvv-intrinsic-doc/pull/41

**Pull request closed without merging**